src/assessments/seo/keywordDensityAssessment.js

```diff
@@ -1,148 +1,265 @@
+				overMaximum: -10,
+				correctDensity: 9,
+				underMinimum: 4,
+			},
+			url: "<a href='https://yoa.st/2pe' target='_blank'>",
+		};
+
+		this.identifier = "keywordDensity";
+		this._config = merge( defaultConfig, config );
 	}
 
-	if ( inRangeEndInclusive( roundedKeywordDensity, 2.5, 3.5 ) ) {
-		score = -10;
+	/**
+	 * Runs the keyword density module, based on this returns an assessment result with score.
+	 *
+	 * @param {Paper} paper The paper to use for the assessment.
+	 * @param {Researcher} researcher The researcher used for calling the research.
+	 * @param {Object} i18n The object used for translations.
+	 *
+	 * @returns {AssessmentResult} The assessment result.
+	 */
+	getResult( paper, researcher, i18n ) {
+		let assessmentResult = new AssessmentResult();
 
-		/* Translators: %1$s expands to the keyword density percentage, %2$d expands to the keyword count,
-		%3$s expands to the maximum keyword density percentage, %4$s expands to a link to a Yoast.com article
-		about keyword density, %5$s expands to the anchor end tag. */
-		text = i18n.dngettext(
-			"js-text-analysis",
-			"The %4$skeyword density%5$s is %1$s, which is over the advised %3$s maximum; the focus keyword was found %2$d time.",
-			"The %4$skeyword density%5$s is %1$s, which is over the advised %3$s maximum; the focus keyword was found %2$d times.",
-			keywordCount
-		);
+		this._keywordCount = researcher.getResearch( "keywordCount" ).count;
 
-		text = i18n.sprintf( text, keywordDensityPercentage, keywordCount, max, url, "</a>" );
-	}
+		this._keywordDensity = researcher.getResearch( "getKeywordDensity" );
 
-	if ( inRangeStartEndInclusive( roundedKeywordDensity, 0.5, 2.5 ) ) {
-		score = 9;
+		const calculatedScore = this.calculateResult( i18n );
+		assessmentResult.setScore( calculatedScore.score );
+		assessmentResult.setText( calculatedScore.resultText );
+		assessmentResult.setHasMarks( this._keywordCount > 0 );
 
-		/* Translators:	%1$s expands to the keyword density percentage,	%2$d expands to the keyword count,
-		%3$s expands to a link to a Yoast.com article about keyword density,
-		%4$s expands to the anchor end tag. */
-		text = i18n.dngettext(
-			"js-text-analysis",
-			"The %3$skeyword density%4$s is %1$s, which is great; the focus keyword was found %2$d time.",
-			"The %3$skeyword density%4$s is %1$s, which is great; the focus keyword was found %2$d times.",
-			keywordCount
-		);
+		return assessmentResult;
+	}
 
-		text = i18n.sprintf( text, keywordDensityPercentage, keywordCount, url, "</a>" );
+	/**
+	 * Checks whether there are no keyword matches in the text.
+	 *
+	 * @returns {boolean} Returns true if the keyword count is 0.
+	 */
+	hasNoMatches() {
+		return this._keywordCount === 0;
 	}
 
-	if ( roundedKeywordDensity < 0.5 && keywordCount === 0 ) {
-		score = 4;
+	/**
+	 * Checks whether there are too few keyword matches in the text. One keyword match is always considered
+	 * as bad, regardless of the density.
```

Comment:
The description needs to be changed - `One keyword match is always considered as bad, regardless of the density. ` only holds true for recalibration.